### DISCUSSION TITLE
1 and 2 electron overlap integrals for fragment

### DISCUSSION THREAD

**Discussion by hh2110** [BoothGroup/Vayesta#34] [Q&A] [answered]
Hi, @a-p25 and I are trying to output the one electron and two electron overlap integrals of a fragment in the fragment's localized orbital basis, after we have performed the fragmentation of a system - does anyone have any initial advice? 

The first thing we thought we'd try and do is look into the `DMETFragment` and `Fragment` classes, and see if they are generated in any calculations, but we couldn't find anything. Our current plan is to try and extract the logic from an OpenQEMIST (http://openqemist.1qbit.com/docs/#) method, which can calculate the overlap integrals from:

- Molecule (pyscf.gto.Mole)
- Mean field (pyscf.scf.RHF)
- Active space, all orbitals in initial SCF calculation
- Localization function (from pyscf.lo.orth.py)
- Bath orbitals
- Number of orbitals in fragment calculation
- Core part of the one-particle RDM

We're now looking at getting these from the `Fragment` object and performing the manipulations on them to get integrals, but was wondering if anyone knows of another way we might be missing. Plus if you think this is useful to the vayesta repo - we'd be happy to create a PR for it.

**Comment by maxnus** [comment]
Hi @hh2110 do you really mean overlap integrals? As in `<i|j>` ? 
Or do you perhaps mean 1- and 2-electron Hamiltonian elements `<i|h|j>` and `<ij|r^-1|kl>`?

**Comment by hh2110** [comment]
I mean `<i|h|j>` and `<ij|r^-1|kl>`

**Comment by maxnus** [comment, marked as answer]
So we have an example here:
https://github.com/BoothGroup/Vayesta/blob/master/examples/ewf/molecules/20-dump-clusters.py
which is specifically for users who would like to use Vayesta to define fragment and bath spaces but then use the resulting spaces and integrals within their own code.

It dumps some cluster information to an HDF5 file - within a different group for each fragment. The 1- and 2-electron integrals, projected onto the cluster (=fragment+bath) space are in the fields `hcore` and `eris`.
There is also `heff`, which is `hcore` plus the Coulomb+exchange effective potential from frozen environment electrons.

If you only want these integrals in the fragment (without bath) space, you would have to do some extra work.
You would first have to form the matrix `r = frac['c_cluster'].T.dot(ovlp).dot(frag['c_frag'])`
which projectes from cluster basis to fragment basis and where `ovlp` is the AO overlap matrix, which you can get from PySCF as `mf.get_ovlp()` or from Vayesta as `emb.get_ovlp()`.

Then you can perform `hcore_frag = np.einsum('ij,ip,jq->pq', frag['hcore'], r, r)` to project the 1e-core Hamiltonian onto the fragment space only. I hope this helps.

**Comment by hh2110** [comment]
Yes, this is perfect thank you!Sorry, I meant I wanted the integrals in the cluster (fragment+bath)
